src/flask_app.py

```diff
@@ -54,3 +68,53 @@ def flask_environment(self) -> dict[str, str]:
                 env[proxy_variable.upper()] = str(proxy_value)
         env.update(self._charm_state.database_uris)
         return env
+
+    def _flask_layer(self) -> ops.pebble.LayerDict:
+        """Generate the pebble layer definition for flask application.
+
+        Returns:
+            The pebble layer definition for flask application.
+        """
+        environment = self._flask_environment()
+        return ops.pebble.LayerDict(
+            services={
+                FLASK_SERVICE_NAME: {
+                    "override": "replace",
+                    "summary": "Flask application service",
+                    "command": shlex.join(self._webserver.command),
+                    "startup": "enabled",
+                    "environment": environment,
+                }
+            },
+        )
+
+    def _update_app_and_unit_status(self, status: ops.StatusBase) -> None:
+        """Update the application and unit status.
+
+        Args:
+            status: the desired application and unit status.
+        """
+        self._charm.unit.status = status
+        if self._charm.unit.is_leader():
+            self._charm.app.status = status
+
+    def restart_flask(self) -> None:
+        """Restart or start the flask service if not started with the latest configuration."""
+        container = self._charm.unit.get_container("flask-app")
+        if not container.can_connect():
+            logger.info("pebble client in the Flask container is not ready")
+            return
+        if not self._charm_state.is_secret_storage_ready:
+            logger.info("secret storage is not initialized, defer config-changed")
```

Comment:
The log msg seems a bit misleading. The event is not deferred

Reply:
Updated, thanks!